internal/utils/state.go

```diff
@@ -0,0 +1,81 @@
+package utils
+
+import (
+	"encoding/json"
+	"errors"
+	"io"
+	"io/fs"
+	"os"
+	"path"
+	"time"
+
+	"github.com/spf13/afero"
+	"github.com/twpayne/go-xdg/v6"
+)
+
+// A State records state.
+type State struct {
+	filename string
+	Upgrade  struct {
+		LastSuccess time.Time
+	}
+}
+
+// NewState returns a state persisted to localAPIFS. If there is no existing
+// state, then an empty state is returned.
+func NewState(localAPIFS afero.Fs) (state *State, err error) {
```

Comment:
Currently the injected FS is called `localAPIFS` because, until now, it was only used for mocking the API (i.e. the tool definitions) during testing. If we use it for the state management as well — which would simplify testing, I guess — we should probably just call it `localFS` or something like that here.

Reply:
To clarify, we should then probably also rename it everywhere else, but that's out of scope for this PR.